### ISSUE TITLE
Store event arguments

### ISSUE THREAD

**Issue by kwalcock** [lum-ai/odinson#113]
This is on my todo list, but it is going to need discussion...

**Comment by kwalcock** [comment]
We were initially storing


Name | Type
-- | --
docBase | Int
docId | Int
label | String
start | Int
end | Int

in the state "table".  If the idea is to make a record of the OdinsonMatches, they also have an Array[NamedCapture], with them being (name, labelOpt, OdinsonMatch), making this a recursive data structure.  SQL is not especially adept at storing such structures.  I'd like to flatten out the potential tree structures in application code just before an OdisonMatch is stored in the state and rebuild the tree after it is retrieved.

The state table will get a few extra columns.  The docBase, docId, and label will all be taken from the root OdinsonMatch and replicated for each child so that a select statement for the parent will collect all the needed children as well and the database need only be accessed once.

Name | Type
-- | --
docBase | Int
docId | Int
label | String
start | Int
end | Int
id | Int
parentId | Int/null
childIndex | Int/null
name | String/null
childLabel | String/null

A root node might look like

Name | Value
-- | --
docBase | 42
docId | 1
label | something
start | 5
end | 10
id | 1
parentId | null
childIndex | null
name | null
childLabel | null

and a child node might look like

Name | Value
-- | --
docBase | 42
docId | 1
label | something
start | 12
end | 14
id | 2
parentId | 1
childIndex | 0
name | myName
childLabel | myLabel

In the SQL version, each State is stored in a separate table, so the state object is able to issue unique integer IDs without the database needing to calculate them and without resorting to GUIDs.  The table gets dropped when the state goes out of scope.

I don't know if the order of Array[NamedCapture] in an OdinsonMatch is significant.  The childIndex can be dropped if not.  

getMatches just now returns an Array[(Int, Int)] to hold the start and end positions.  That probably needs to be turned into an array of OdinsonMatch.  I'm not yet sure how that needs to be integrated into the caller StateSpans.advance.

There are lots of names and optional labels floating around.  I hope they are used appropriately.

Can any query actually access the children?

I can hopefully make the state queries for getDocIds and getMatches return iterators so that the results do not need to be stored.  That will be an optimization after the functionality is more complete.

**Comment by kwalcock** [comment]
At the meeting the other day I was made aware that in the State should be stored everything needed to create a new Mention.  The constructor for a mention,
```scala
  def newMention(
    odinsonMatch: OdinsonMatch,
    label: Option[String],
    luceneDocId: Int,
    luceneSegmentDocId: Int,
    luceneSegmentDocBase: Int,
    docId: String,
    sentenceId: String,
    foundBy: String,
    arguments: Map[String, Array[Mention]],
  ): Mention

sealed trait OdinsonMatch {

  def start: Int
  def end: Int
  def namedCaptures: Array[NamedCapture]
}

case class NamedCapture(name: String, label: Option[String], capturedMatch: OdinsonMatch)
```
So, foundBy, docId (the String), and an additional Lucene integer seem to need to be stored.

**Comment by BeckySharp** [comment]
@kwalcock -- is this one done? or is it waiting on the idea of reconstituting mentions? (or none of the above?)

**Comment by kwalcock** [comment]
Yes, finished.  I forgot to assign myself so that it then didn't show up in my search.
